sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -123,15 +123,16 @@ public static Write write() {
   public abstract static class Write extends PTransform<PCollection<Mutation>, PDone> {
 
     @Nullable
-    abstract String getProjectId();
+    abstract ValueProvider<String> getProjectId();
 
     @Nullable
-    abstract String getInstanceId();
+    abstract ValueProvider<String> getInstanceId();
 
     @Nullable
-    abstract String getDatabaseId();
+    abstract ValueProvider<String> getDatabaseId();
 
-    abstract long getBatchSizeBytes();
+    @Nullable
+    abstract ValueProvider<Long> getBatchSizeBytes();
```

Comment:
Speaking of... Do you expect users to have to configure this in practice? Can SpannerIO choose a reasonable value and shield users from the temptation of "optimizing" it (which might fall out of date as Spanner implementation changes, or might be simply chosen misguidedly)?